File: src/day6.py
import unittest


def part_one() -> int:
    with open("src/day6_input") as f:
        return find_marker(f.read().strip(), 4)
    return -1


def part_two() -> int:
    with open("src/day6_input") as f:
        return find_marker(f.read().strip(), 14)
    return -1


def find_marker(text: str, marker_len: int) -> int:
    for i in range(len(text) - marker_len + 1):
        if len(set(text[i : i + marker_len])) == marker_len:
            return i + marker_len
    return -1


class TestDevice(unittest.TestCase):
    def test_part_one(self):
        self.assertEqual(find_marker("mjqjpqmgbljsphdztnvjfqwrcgsmlb", 4), 7)
        self.assertEqual(find_marker("bvwbjplbgvbhsrlpgdmjqwftvncz", 4), 5)
        self.assertEqual(find_marker("nppdvjthqldpwncqszvftbrmjlhg", 4), 6)
        self.assertEqual(find_marker("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", 4), 10)
        self.assertEqual(find_marker("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", 4), 11)

    def test_part_two(self):
        self.assertEqual(find_marker("mjqjpqmgbljsphdztnvjfqwrcgsmlb", 14), 19)
        self.assertEqual(find_marker("bvwbjplbgvbhsrlpgdmjqwftvncz", 14), 23)
        self.assertEqual(find_marker("nppdvjthqldpwncqszvftbrmjlhg", 14), 23)
        self.assertEqual(find_marker("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", 14), 29)
        self.assertEqual(find_marker("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", 14), 26)


if __name__ == "__main__":
    print(part_two())

File: src/day6.rs
use std::collections::HashSet;
use std::fs;

pub fn resolve(part: u32) {
    let lines = fs::read_to_string("src/day6_input").expect("Can't read file");

    match part {
        1 => println!("{}", part_one(lines)),
        2 => println!("{}", part_two(lines)),
        _ => (),
    }
}

fn find_marker(text: String, mark_len: usize) -> usize {
    for i in 0..text.len() - mark_len + 1 {
        let s: HashSet<char> = HashSet::from_iter(text[i..i + mark_len].chars());
        if s.len() == mark_len {
            return i + mark_len;
        }
    }
    0
}

fn part_one(text: String) -> usize {
    find_marker(text, 4)
}

fn part_two(text: String) -> usize {
    find_marker(text, 14)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        assert_eq!(part_one(String::from("mjqjpqmgbljsphdztnvjfqwrcgsmlb")), 7);
        assert_eq!(part_one(String::from("bvwbjplbgvbhsrlpgdmjqwftvncz")), 5);
        assert_eq!(part_one(String::from("nppdvjthqldpwncqszvftbrmjlhg")), 6);
        assert_eq!(
            part_one(String::from("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg")),
            10
        );
        assert_eq!(
            part_one(String::from("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw")),
            11
        );
    }

    #[test]
    fn test_part_two() {
        assert_eq!(part_two(String::from("mjqjpqmgbljsphdztnvjfqwrcgsmlb")), 19);
        assert_eq!(part_two(String::from("bvwbjplbgvbhsrlpgdmjqwftvncz")), 23);
        assert_eq!(part_two(String::from("nppdvjthqldpwncqszvftbrmjlhg")), 23);
        assert_eq!(
            part_two(String::from("nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg")),
            29
        );
        assert_eq!(
            part_two(String::from("zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw")),
            26
        );
    }
}
